File: Backend/Events/Converters/NpcConverter.cs
using Backend.Domain.Models;
using Backend.Domain.Models.Npcs;
using Backend.Events.DTO.Npcs;

namespace Backend.Events.Converters;

public static class NpcConverter
{
    public static NpcDTO ToDTO(Npc npc)
    {
        return new NpcDTO
        {
            Id = npc.Id,
            DigimonBattles = npc.DigimonBattles.Select(ToBattleDTO).ToList(),
            CardBattles = npc.CardBattles.Select(ToBattleDTO).ToList(),
        };
    }

    public static NpcBattleDTO ToBattleDTO(NpcBattle battle)
    {
        return new NpcBattleDTO
        {
            Id = battle.Id,
            Value = battle.Value,
        };
    }
}

File: Frontend/src/events/converters/npcs/npc.converter.ts
import type { NpcBattleDTO } from "@/events/dto/npcs/npc-battle.dto";
import type { NpcDTO } from "@/events/dto/npcs/npc.dto";
import type { Npc, NpcBattle } from "@/models";

export class NpcConverter {
  public static convert(npcDto: NpcDTO): Npc {
    return {
      id: npcDto.id,
      digimonBattles: (npcDto.digimonBattles ?? []).map((battleDto) => {
        return this.convertBattle(battleDto);
      }),
    };
  }

  public static convertBattle(battleDto: NpcBattleDTO): NpcBattle {
    return {
      id: battleDto.id,
      completed: battleDto.value !== undefined && battleDto.value !== 0,
    };
  }
}
